mrfApp/src/drvLinuxEvr.c

```diff
@@ -198,69 +214,147 @@ void ErIrqHandler(int fd, int flags)
+			if (fe->EventCode == 1)				                     \
+			    pCard->havefid = 1;                                              \
+			if (pCard->DevEventFunc != NULL)		                     \
+			    (*pCard->DevEventFunc)(pCard, fe->EventCode, fe->TimestampHigh); \
+		    }							                     \
+		}
+
+#define SKIP_DATA()                                                                          \
+		while (drp < pEq->dwp) {					             \
+		    int idx = drp & (MAX_EVR_DBQ2-1);			                     \
+		    int databuf_sts = pEq->dbq2[idx].status;		                     \
+		    u32 *dd = pEq->dbq2[idx].data;				             \
+		    if (databuf_sts & (1<<C_EVR_DATABUF_CHECKSUM)) {                         \
+			if (pCard->DevErrorFunc != NULL)		                     \
+			    (*pCard->DevErrorFunc)(pCard, ERROR_DBUF_CHECKSUM);              \
+			drp++;					                             \
+		    } else if (dd[8] == lastnsec) {				             \
+			drp++;                                                               \
+	            } else						                     \
+			break;						                     \
+		}
+
+#define HANDLE_DATA()                                                                        \
+		{							                     \
+		    int idx = drp & (MAX_EVR_DBQ2-1);			                     \
+		    int databuf_sts = pEq->dbq2[idx].status;		                     \
+		    u32 *dd = pEq->dbq2[idx].data;			                     \
+		    if (pCard->DevDBuffFunc != NULL && pCard->havefid) {		     \
+			pCard->DBuffSize = (databuf_sts &		                     \
+		                            ((1<<(C_EVR_DATABUF_SIZEHIGH+1))-1));	     \
+			memcpy(pCard->DataBuffer, dd, pCard->DBuffSize);                     \
+			(*pCard->DevDBuffFunc)(pCard, pCard->DBuffSize,	pCard->DataBuffer);  \
+	            }								             \
+		    lastnsec = dd[8];					                     \
+		    drp++;						                     \
 		}
 
+		/*
+		 * If we are far behind events, try to catch up, but flag an error!
+		 */
+		if (pEq->ewp - erp > MAX_EVR_EVTQ / 2) {
+		    erp = pEq->ewp - MAX_EVR_EVTQ / 2;
+		    flags |= EVR_IRQFLAG_FIFOFULL;
+		    printf("EVENT FIFO FULL!!\n");
+
+		    /* Now, skip forward to next fiducial event! */
+		    while (erp < pEq->ewp && pEq->evtq[erp & (MAX_EVR_EVTQ-1)].EventCode != 1)
+			erp++;
+		    if (erp < pEq->ewp) /* Skip the fiducial itself! */
+			erp++;
+                }
+
+		SKIP_DATA();   /* Skip if we have duplicates or errors! */
+
+		SET_NEXT_EFID();
+		SET_NEXT_DFID();
+
+		while (erp < pEq->ewp || drp < pEq->dwp) {
+		    /*
+		     * Sigh.  Our macro assumes valid fiducials.  So we need to write the
+		     * comparison carefully.
+		     */
+		    if (nextEfid == FID_INVALID || 
+			(nextDfid != FID_INVALID && FID_GT(nextEfid, nextDfid))) {
+			/* Process a data buffer! */
+			HANDLE_DATA();
+			SKIP_DATA();
+			usleep(1000);
```

Comment:
In an ideal world... we'd get a data buffer, interrupt for that, then 2ms later we'd get some events, interrupt for them (maybe more than once, but... eventually processing all of them), then 0.7ms later, we'd get another data buffer.  Each interrupt would either handle events or data buffers, but not both.

That's what the original code was written to expect.  Unfortunately... it doesn't always happen that way.  Usually, yes... but not always.

If it *does* happen this way, life is good, one of the next fids will be set, the other will be invalid, and we will process things in the proper order.

The issue is that sometimes, Linux acts out and delays processing for 10ms or more.  I've seen as many as *4* data buffers queued up to be processed in one call of the handler, with a similar number of events.  Not sure *why* things get so bad... but unless we can solve *this*, the code needs to deal with things piling up.

The 1ms delay is actually important here... because if we don't give the lower priority parts of the code a chance to run and catch up, we are just pushing the pile up further down the line.

(If we get to HANDLE_DATA, there *must* be an unprocessed item at the head of the data queue, so we know there is data there.  If *more* data arrives... after we handle and skip duplicates, we set the "next data fid" again, and that will find the newly arrived data element.)